rfcs/0159-private-task-logs.md

```diff
@@ -0,0 +1,66 @@
+
+The Taskcluster UI, and other consumers of task logs such as Treeherder, incorporate this convention and look for logs at that artifact name.
+
+Workers that cannot do "live logging" simply create an S3 artifact with this name.
+Workers that *can* do live logging use the following process.
+At the beginning of the task, they create a Reference artifact named `public/logs/live.log` containing a URL that will connect to the worker (usually via [Websocktunnel](https://github.com/taskcluster/taskcluster/tree/master/tools/websocktunnel#readme)) to get log lines as they are produced.
+When the task is complete, they create a new S3 artifact containing the whole log, named `public/logs/live_backing.log`.
+They then modify the URL in the `public/logs/live.log` artifact (using a [special case in the `createTask` API](https://docs.taskcluster.net/docs/reference/platform/queue/api#createArtifact)) to point to the URL for the `getArtifact` method for the `public/logs/live_backing.log` artifact (`<rootUrl>/api/queue/v1/task/<taskId>/runs/<runId>/artifacts/public/logs/live_backing.log`).
+
+Reference artifacts correspond to HTTP 303 redirects, and HTTP clients do not re-authenticate when following a redirect.
+So the fetch of `live_backing.log` is performed without Taskcluster credentials, even if the initial call to `getArtifact` used credentials.
+If `live_backing.log` is always public, this is not a problem, but if it is private then this may result in additional complexity for clients wishing to access task logs.
+
+## New Log Implementation
+
+The updated convention is, briefly, to make the artifact name containing logs configurable on a per-task basis, so that users may choose a `public/` or non-public name; and to introduce a new artifact type that will avoid the issues around authentication and redirects.
+
+The convention for task-related conventions is to put information in `task.extra`, using sub-objects to group related configuration.
+In the new design, the log artiact name is found at `task.extra.logs.taskLogArtifact`, with a default value of `public/logs/live.log` for compatibility.
+Workers will be modified to consult this field before creating task log artifacts.
+UIs and other log consumers will be modified to consult this field before fetching task logs.
+
+To support live logging, a new queue artifact type, "LiveLog", is added.
+This type is similar to the existing Reference type, in that it contains a URL to which it redirects.
+However, its URL cannot be modified, but it can be replaced with an artifact of another type.
+
+Workers create a LiveLog artifact when a task begins, and replace that with a "normal" artifact (such as S3) when the task completes.
+The URL in the LiveLog artifact should contain enough entropy that it cannot be guessed, preventing anyone who cannot read the LiveLog artifact from connecting to the worker to download logs.
```

Comment:
Reading the `New Log Implementation` section (and this last paragraph in particular), it is unclear to me what the protections for this new "non-public" task log will be. Will a user be required to have the needed scopes to fetch these logs? How will that be implemented?

Reply:
Yes, and that's already implemented -- a caller needs `queue:get-artifact:<name>` to retrieve an artifact *except* when the artifact name begins with `public/` in which case it's public :)

Reply:
Ok totally, then why is `The URL in the LiveLog artifact should contain enough entropy that it cannot be guessed, preventing anyone who cannot read the LiveLog artifact from connecting to the worker to download logs.` a thing? Is the problem here that a user who has scopes to interact with a worker directly but not download logs could just connect to the worker directly and download them?

Reply:
Currently, the URL for a live-log points directly (or via websocktunnel) to the worker, and the protection against anyone reading that data is that a portion of the URL contains a random value that we presume can't be guessed (it's a slugid, so ~128 bits from a CSPRNG).  There's also a limited time window during which that data is available (the time during which the task is executing).  The `queue:get-artifact:<name>` scope would prevent unauthorized access to the URL.

It's not hard to imagine ways to do that better -- workers accepting TC credentials, websocktunnel doing authentication, etc. -- but it would be harder to implement those ideas.  I'd like to leave them out-of-scope for this RFC, if possible.

Reply:
@djmitche Could we schedule some time to chat about this over zoom? I feel like I may be missing some context. If we are talking about utilizing private logs for things like critical security bugs logs and the only protection is a random value in the URL, that sounds like a big incident just waiting to happen. It just takes one bug that causes the URL to stay available (or something else) and we've got a big 'ol incident on our hands.

Reply:
We can.  But basically that's what I'm proposing.  Maybe for such cases we would disable live logging -- the access control for the log once the task is complete does not rely on a secret URL.